src/metrics.py

```diff
@@ -68,32 +73,61 @@ def _get_event_name(event: Event) -> str:
 
 
 def issue_event(event: Event) -> None:
-    """Transmit an event to Promtail.
+    """Issue a metric event.
+
+    The metric event is logged to the metrics log.
 
     Args:
         event: The metric event to log.
     Raises:
-        requests.HTTPError: If the HTTP request to Promtail fails.
+        OSError: If an error occurs while writing the metrics log.
     """
-    if not promtail.is_running():
-        logger.warning("Promtail is not running, skipping event transmission")
-        return
-
     event_dict = event.dict()
-    event_dict["event"] = _get_event_name(event)
-
-    resp = requests.post(
-        PROMTAIL_PUSH_API_URL,
-        json={
-            "streams": [
-                {
-                    "stream": {"job": "metrics"},
-                    "values": [
-                        [str(time.time_ns()), json.dumps(event_dict)],
-                    ],
-                }
-            ]
-        },
-        timeout=5,
+    event_name = _get_event_name(event)
+    event_dict["event"] = event_name
+
+    with METRICS_LOG_PATH.open(mode="a", encoding="utf-8") as metrics_file:
+        metrics_file.write(f"{json.dumps(event_dict)}\n")
+
+
+def _enable_logrotate() -> None:
+    """Enable and start the logrotate timer if it is not active.
+
+    Raises:
+        SubprocessError: If the logrotate.timer cannot be enabled and started.
+    """
+    execute_command([SYSTEMCTL_PATH, "enable", LOG_ROTATE_TIMER_SYSTEMD_SERVICE], check_exit=True)
+
+    _, retcode = execute_command(
+        [SYSTEMCTL_PATH, "is-active", "--quiet", LOG_ROTATE_TIMER_SYSTEMD_SERVICE]
     )
-    resp.raise_for_status()
+    if retcode != 0:
+        execute_command(
+            [SYSTEMCTL_PATH, "start", LOG_ROTATE_TIMER_SYSTEMD_SERVICE], check_exit=True
+        )
+
+
+def _configure_logrotate() -> None:
+    """Configure logrotate for the metrics log."""
+    # Set rotate to 0 to not keep the old metrics log file to avoid sending the
+    # metrics to Loki twice, which may happen if there is a corrupt log scrape configuration.
+    LOGROTATE_CONFIG.write_text(
+        f"""{str(METRICS_LOG_PATH)} {{
+    rotate 0
+    missingok
+    notifempty
+    create
+}}
+""",
+        encoding="utf-8",
+    )
+
+
+def setup_logrotate():
+    """Configure logrotate for the metrics log.
+
+    Raises:
+        SubprocessError: If the logrotate.timer cannot be enabled.
+    """
+    _enable_logrotate()
+    _configure_logrotate()
```

Comment:
Shouldn't handle to potential exception raised here ?

Reply:
An expected exception would be a `SubprocessError` which is passed through and handled by the caller in `charm.py`. 
I prefer not to catch it here in order to avoid that logrotate will never run if there is a systematic issue (e.g. `logrotate` is not installed). What do you think?

Reply:
Updated the code to throw a custom exception.